internal/controller/supervisorconfig/oidcclientwatcher/oidc_client_watcher.go

```diff
@@ -0,0 +1,317 @@
+		})
+	case allowedScopesContains(oidcClient, requestAudienceScopeName) &&
+		(!allowedScopesContains(oidcClient, usernameScopeName) || !allowedScopesContains(oidcClient, groupsScopeName)):
+		conditions = append(conditions, &v1alpha1.Condition{
+			Type:   allowedScopesValid,
+			Status: v1alpha1.ConditionFalse,
+			Reason: reasonMissingRequiredValue,
+			Message: fmt.Sprintf("%q and %q must be included in %q when %q is included in %q",
+				usernameScopeName, groupsScopeName, allowedScopesFieldName, requestAudienceScopeName, allowedScopesFieldName),
+		})
+	case allowedGrantTypesContains(oidcClient, tokenExchangeGrantTypeName) && !allowedScopesContains(oidcClient, requestAudienceScopeName):
+		conditions = append(conditions, &v1alpha1.Condition{
+			Type:   allowedScopesValid,
+			Status: v1alpha1.ConditionFalse,
+			Reason: reasonMissingRequiredValue,
+			Message: fmt.Sprintf("%q must be included in %q when %q is included in %q",
+				requestAudienceScopeName, allowedScopesFieldName, tokenExchangeGrantTypeName, allowedGrantTypesFieldName),
+		})
+	default:
+		conditions = append(conditions, &v1alpha1.Condition{
+			Type:    allowedScopesValid,
+			Status:  v1alpha1.ConditionTrue,
+			Reason:  reasonSuccess,
+			Message: fmt.Sprintf("%q is valid", allowedScopesFieldName),
+		})
+	}
+	return conditions
+}
+
+// validateAllowedGrantTypes checks if allowedGrantTypes is valid on the OIDCClient.
+func validateAllowedGrantTypes(oidcClient *v1alpha1.OIDCClient, conditions []*v1alpha1.Condition) []*v1alpha1.Condition {
+	switch {
+	case !allowedGrantTypesContains(oidcClient, authorizationCodeGrantTypeName):
+		conditions = append(conditions, &v1alpha1.Condition{
+			Type:   allowedGrantTypesValid,
+			Status: v1alpha1.ConditionFalse,
+			Reason: reasonMissingRequiredValue,
+			Message: fmt.Sprintf("%q must always be included in %q",
+				authorizationCodeGrantTypeName, allowedGrantTypesFieldName),
+		})
+	case allowedScopesContains(oidcClient, offlineAccessScopeName) && !allowedGrantTypesContains(oidcClient, refreshTokenGrantTypeName):
+		conditions = append(conditions, &v1alpha1.Condition{
+			Type:   allowedGrantTypesValid,
+			Status: v1alpha1.ConditionFalse,
+			Reason: reasonMissingRequiredValue,
+			Message: fmt.Sprintf("%q must be included in %q when %q is included in %q",
+				refreshTokenGrantTypeName, allowedGrantTypesFieldName, offlineAccessScopeName, allowedScopesFieldName),
+		})
+	case allowedScopesContains(oidcClient, requestAudienceScopeName) && !allowedGrantTypesContains(oidcClient, tokenExchangeGrantTypeName):
+		conditions = append(conditions, &v1alpha1.Condition{
+			Type:   allowedGrantTypesValid,
+			Status: v1alpha1.ConditionFalse,
+			Reason: reasonMissingRequiredValue,
+			Message: fmt.Sprintf("%q must be included in %q when %q is included in %q",
+				tokenExchangeGrantTypeName, allowedGrantTypesFieldName, requestAudienceScopeName, allowedScopesFieldName),
+		})
```

Comment:
I would expect to see all errors on the status, not just the first matching case.